pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -93,99 +96,151 @@ func requeueAfterDuration(shoot *gardencorev1beta1.Shoot) (time.Duration, time.T
+
+		reasonForKubernetesUpdate = append(reasonForKubernetesUpdate, reasonForWorkerPoolKubernetesUpdate...)
+	}
+	operations = append(operations, reasonForKubernetesUpdate...)
+
+	operation := maintainOperation(maintainedShoot)
+	if operation != "" {
+		operations = append(operations, fmt.Sprintf("Added %q operation annotation", operation))
+	}
+
+	if len(operations) > 0 {
+		patch := client.MergeFrom(shoot.DeepCopy())
+		shoot.Status.LastMaintenance = &gardencorev1beta1.LastMaintenance{
+			Description:   strings.Join(operations, ", "),
+			TriggeredTime: metav1.Time{Time: r.Clock.Now()},
+			State:         gardencorev1beta1.LastOperationStateProcessing,
+		}
+
+		// First dry run the update call to check if it can be executed successfully.
+		// If not shoot maintenance is marked as failed and is retried only in
+		// next maintenance window.
+		if err := r.Client.Update(ctx, maintainedShoot, &client.UpdateOptions{
+			DryRun: []string{metav1.DryRunAll},
+		}); err != nil {
+			// If shoot maintenance is triggerd by `gardener.cloud/operation=maintain` annotation and if it fails in dry run
+			// maintain operation annotation need to be removed so that if reason for failure is fixed and maintenance is triggerd
+			// again using maintain operation annotation then it should not fail with the reason that annotation is already present
+			if hasMaintainNowAnnotation(shoot) {
+				delete(shoot.Annotations, v1beta1constants.GardenerOperation)
+			}
+			shoot.Status.LastMaintenance.State = gardencorev1beta1.LastOperationStateFailed
+			shoot.Status.LastMaintenance.FailureReason = pointer.String(fmt.Sprintf("Maintenance failed: %s", err.Error()))
+			if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
+				return err
+			}
+
+			log.Info("Shoot maintenance failed", "reason", err)
+			return nil
+		}
+
+		if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
+			return err
+		}
 	}
 
-	maintainOperation(shoot)
+	// update shoot spec changes in maintenance call
+	shoot.Spec = *maintainedShoot.Spec.DeepCopy()
+	_ = maintainOperation(shoot)
 	maintainTasks(shoot, r.Config)
 
 	// try to maintain shoot, but don't retry on conflict, because a conflict means that we potentially operated on stale
 	// data (e.g. when calculating the updated k8s version), so rather return error and backoff
-	if err := gardenClient.Update(ctx, shoot); err != nil {
+	if err := r.Client.Update(ctx, shoot); err != nil {
 		r.Recorder.Event(shoot, corev1.EventTypeWarning, gardencorev1beta1.ShootMaintenanceFailed, err.Error())
 		return err
 	}
 
+	if shoot.Status.LastMaintenance != nil && shoot.Status.LastMaintenance.State == gardencorev1beta1.LastOperationStateProcessing {
+		patch := client.MergeFrom(shoot.DeepCopy())
+		shoot.Status.LastMaintenance.State = gardencorev1beta1.LastOperationStateSucceeded
+		if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
```

Comment:
Do we have to set `FailureReason` to `nil` here in case it was set before and the issue preventing maintenance got fixed meanwhile?

Reply:
No, For a reconciliation to reach this point, it should succeed in previous steps, i.e. for the case, if the failure reason is set, the next reconciliation will reach this point with shoot status having a new last maintenance value in which the failure reason will be nil.